### PR TITLE
Arc 2119 refactor backfill deployment

### PR DESCRIPTION
**Pull request by gxueatlassian**
**What's in this PR?**
This PR is refactor our backfill logic for deployments.

1. Save the deployment infos in dynamo db when fetched from GraphQL
2. Look up last success from dynamo first, and fallback to current behaviour if not found.

**Why**
1. Our backfill logic for deployment is incorrect. Each time processing a deployment, it try to call rest api to fetch last 100 success deployment from today. Which causes it to compare itself to "future" deployment. By saving deployments in dynamo during the graphql fetch, we can look it up, and hopefully find the correct "last" success deployment. `NOTE`: This is PR is **NOT** to make the logic 100% correct. As 1) it is hard to guarantee that. Lots of edge cases. 2) This PR however will "significantly", <-- I hope, improve the accuracy of finding a last success deployments.
2. Our backfill processing for deployments is SLOOOOOWWW (unefficient to be specific). It first fetch 100 from graphql, and each of them, will call a rest api to 100 latest deployments of that env, which in turns each call status api to find a success one. So the time complexity of this logic if O(n^3) 😱 By using dynamo, I will say most of the time we can find "last successful deployment" in the dynamodb. But this need to be verified in the metrics.

**Added feature flags**
use-dynamodb-for-deployment-backfill

**Affected issues**  
ARC-2119

**How has this been tested?**  
Unit tested
stg tested

**Whats Next?**
Depends on the metrics and how far we want to go, we can:
1. Refactor the process furthur more to split backfill deployment into separate two stage. Stage 1 scan all deployment in whole history (and save in dynamo), stage 2 process each of them. <----- however this "a little bit" conflicting with our incremental/stagger backfill. Need more thinking.
2. Right now in this PR app fetch an extra page to try to improve efficiency. But the cursor of this extra page is not saved in RepoSyncStates(deploymentCursor) <--- we can make it cursor 1) json 2) compatible with existing 3) record both the scaning/extra fetch and the current processing cursor. <------- however, this seems much more effort, and I'd like to have this PR merged in first, look at the metrics and make decision later.

TLDR; Thank you for reading to end. Such a loooon desc.

### CODE REVIEW

File: src/sync/deployment.test.ts [hunk 1]
```diff
@@ -30,6 +37,187 @@ describe("sync/deployments", () => {
...
+						.reply(200, {
+							commit: {
+								author: { name: "random", email: "random", date: new Date() },
+								message: `commit message [JIRA-${item._seq}]`
+							}, html_url: "random"
+						});
+				});
+			});
+		};
+		const expectDeploymentEntryInDB = async (deployments) => {
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:130]: nit: can we add a new line between each variable declaration and each test? There's a lot of code here and the readability really drops when everything is clumped together without any spaces in between.

Comment (gxueatlassian) [reply src/sync/deployment.test.ts:210]: Have add spaces in between, shuffle them a round, put some functions inline.
Hopefully it can be easier to read.

File: src/sync/deployment.test.ts [hunk 2]
```diff
@@ -30,6 +37,221 @@ describe("sync/deployments", () => {
...
+
+					expectEdgesAndPayloadMatchToDeploymentCommits(result, [deployments[3], deployments[2]]);
+				});
+			});
+
+			describe("for existing deployment cursor", () => {
+				// eslint-disable-next-line jest/expect-expect
+				it("should fetch deployments from existing cursor", async () => {
+
+					const deployments = createDeploymentEntities(4);
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:128]: Can we convert the 'magic number' of 4 to a variable so it's more explicit and gives more context? e.g.

```
 const deploymentCount = 4;
  const deployments = createDeploymentEntities(deploymentCount);
```

File: src/sync/deployment.test.ts [hunk 3]
```diff
@@ -30,6 +37,221 @@ describe("sync/deployments", () => {
...
+								message: `commit message [JIRA-${item._seq}]`
+							}, html_url: "random"
+						});
+				});
+			});
+		};
+
+		const expectDeploymentEntryInDB = async (deployments) => {
+			for (const deployment of deployments) {
+				const key = createHashWithoutSharedSecret(`ghurl_${gitHubCloudConfig.baseUrl}_repo_${deployment.node.repository.id}_env_${deployment.node.environment}`);
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:226]: To tighten this up a little in terms of readability can we throw in some destructuring at the start

```
const {
    repository: { id: repoId },
    environment,
    commitOid
  } = deployment.node;
```

File: src/sync/deployment.test.ts [hunk 4]
```diff
@@ -30,6 +37,221 @@ describe("sync/deployments", () => {
...
+				const successStatusDate = deployment.node.statuses.nodes.find(n=>n.state === "SUCCESS")?.updatedAt;
+				const result = await ddb.getItem({
+					TableName: envVars.DYNAMO_DEPLOYMENT_HISTORY_CACHE_TABLE_NAME,
+					Key: {
+						"Id": { "S": key },
+						"CreatedAt": { "N": String(new Date(successStatusDate).getTime()) }
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:232]: Could probably move this out into a variable as well? `const createdAtTimestamp = successStatusDate && new Date(successStatusDate).getTime();` Then `"CreatedAt": { "N": String(createdAtTimestamp) }`?

File: src/sync/deployment.ts [hunk 5]
```diff
@@ -49,12 +72,39 @@ const getTransformedDeployments = async (deployments, gitHubInstallationClient:
...
-		.filter(deployment => !!deployment)
+	return (transformedDeployments
+		.filter(deployment => !!deployment) as JiraDeploymentBulkSubmitData[])
 		.map(deployment => deployment.deployments)
 		.flat();
 };
 
+const saveDeploymentsForLaterUse = async (deployments: FetchDeploymentResponse["deployments"], gitHubBaseUrl: string, logger: Logger) => {
+	try {
+		const successDeployments: FetchDeploymentResponse["deployments"] = deployments.filter(d => (d.statuses?.nodes.some(n => n.state === "SUCCESS")));
```
Comment (rachellerathbone) [inline src/sync/deployment.ts:83]: nit: same thing here with spacing. There's a lot of lines of code here and it gets really hard to separate the logic when it's all crammed together.

File: src/sync/deployment.ts [hunk 6]
```diff
@@ -49,12 +72,39 @@ const getTransformedDeployments = async (deployments, gitHubInstallationClient:
...
+		logger.info({ deploymentsCount: deployments.length, successDeploymentsCount: successDeployments.length }, "Try to save deployments for later use");
+		const result = await Promise.allSettled(successDeployments.map(dep => {
+			const successStatusDate = dep.statuses?.nodes.find(n=>n.state === "SUCCESS")?.updatedAt;
+			if (!successStatusDate) {
+				logger.warn("Should find a success status date, but found none");
+				throw "Save failure";
```
Comment (rachellerathbone) [inline src/sync/deployment.ts:89]: Can we change this to throw new Error("Save failure")? It's generally better to throw an instance of the Error object rather than throwing a simple string as the Error object provides more info about the error. This will provide more context about what went wrong. Also, the error object will capture the stack trace and we can leverage it for better error handling.

File: src/sync/deployment.test.ts [hunk 7]
```diff
@@ -30,6 +37,221 @@ describe("sync/deployments", () => {
...
+			});
+		};
+
+		const nockFetchingDeploymentgPagesGraphQL = (query, cursor, deployments) => {
+			githubNock.post("/graphql", { query, variables: { owner: repoSyncState.repoOwner, repo: repoSyncState.repoName, per_page: PAGE_SIZE__TWO_ITEMS, cursor } })
+				.query(true).reply(200, { "data": { "repository": { "deployments": { "edges": deployments } } } });
+		};
+
+		const nockDeploymentListingApi = (deployments, repeatTimes) => {
+			"*".repeat(repeatTimes).split("").forEach(() => {
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:176]: Can we change `"*".repeat(repeatTimes).split("")` to `Array(repeatTimes).fill()` as the latter is more performant. The .fill() method will create an array directly with the desired length, whereas `repeat()` followed by `split()` involves additional string operations that can be slower. Might not matter too much coz this is just a test but maybe worth considering ¯\_(ツ)_/¯

File: src/sync/deployment.test.ts [hunk 8]
```diff
@@ -30,6 +37,221 @@ describe("sync/deployments", () => {
...
+				githubNock.get(`/repos/test-repo-owner/test-repo-name/deployments?environment=prod&per_page=10`)
+					.reply(200, deployments.map((item, idx) => ({
+						"id": item.node.databaseId,
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:179]: Can we remove the unnecessary quotations marks around the object property names? These are all valid identifiers and should keep this consistent with the rest of our codebase.

File: src/sync/deployment.test.ts (hunk 4 above)
Comment (rachellerathbone) [reply src/sync/deployment.test.ts:232]: Same with the quotation marks too.

File: src/sync/deployment.test.ts (hunk 2 above)
Comment (gxueatlassian) [reply src/sync/deployment.test.ts]: fixed.

File: src/sync/deployment.test.ts (hunk 3 above)
Comment (gxueatlassian) [reply src/sync/deployment.test.ts]: fixed (to my best effort 🤷 )

File: src/sync/deployment.test.ts (hunk 4 above)
Comment (gxueatlassian) [reply src/sync/deployment.test.ts]: fixed.

File: src/sync/deployment.ts (hunk 5 above)
Comment (gxueatlassian) [reply src/sync/deployment.ts:84]: fixed (to my best effort 🤷 )

File: src/sync/deployment.ts (hunk 6 above)
Comment (gxueatlassian) [reply src/sync/deployment.ts]: Good call!!
Fixed.

File: src/sync/deployment.test.ts (hunk 7 above)
Comment (gxueatlassian) [reply src/sync/deployment.test.ts:179]: Thanks for picking up this. I tried that no luck though. Seems the `Array(number)` create an something empty (although with `length` equal to that number), so when do `Array(number).forEach()` it won't actually loop as it consider it is still an empty array.

Leave this unchanged.

File: src/sync/deployment.test.ts (hunk 8 above)
Comment (gxueatlassian) [reply src/sync/deployment.test.ts]: Fixed.

File: src/sync/deployment.ts (hunk 5 above)
Comment (rachellerathbone) [reply src/sync/deployment.ts:84]: Love your work sir.

File: src/sync/deployment.test.ts [hunk 9]
```diff
@@ -30,6 +37,225 @@ describe("sync/deployments", () => {
...
+
+		describe("when ff is off", () => {
+
+			beforeEach(() => {
+				when(booleanFlag).calledWith(BooleanFlags.USE_DYNAMODB_FOR_DEPLOYMENT_BACKFILL, jiraHost).mockResolvedValue(false);
+			});
+
+			describe("for empty demployment cursor", () => {
+				// eslint-disable-next-line jest/expect-expect
+				it("should fetch deployments from begining", async () => {
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:107]: nit: typo -> beginning

File: src/sync/deployment.test.ts (hunk 7 above)
Comment (rachellerathbone) [reply src/sync/deployment.test.ts:179]: actually, `Array.from({ length: repeatTimes }).forEach(() => {` should do the trick

File: src/sync/deployment.test.ts [hunk 10]
```diff
@@ -30,6 +37,221 @@ describe("sync/deployments", () => {
...
+		//--helper funcs--
+		const msgPayload = () => ({
+			jiraHost,
+			installationId: DatabaseStateCreator.GITHUB_INSTALLATION_ID,
+			commitsFromDate: "2023-01-01T00:00:00Z"
+		});
+
+		//size up to 9 entities
+		const createDeploymentEntities = (size: number) => {
+			return "*".repeat(size).split("").map((_, idx)=> {
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts]: We can improve performance here too: `return Array.from({ length: size }, (_, idx) => {`

File: src/sync/deployment.test.ts [hunk 11]
```diff
@@ -30,6 +37,225 @@ describe("sync/deployments", () => {
...
+					.reply(200, deployments.map((item, idx) => ({
+						id: item.node.databaseId,
+						sha: item.node.commitOid,
+						ref: "random",
+						task: `task for deployment ${idx + 1}`,
+						payload: {},
+						original_environment: item.node.environment,
+						environment: item.node.environment,
+						description: `description for deployment ${idx + 1}`,
+						creato: { login: "test-repo-owner", id: 1, type: "User" },
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:190]: a couple of typos after removing the quotes:

- creato -> creator
- production_environmen -> production_environment

File: src/sync/deployment.test.ts [hunk 12]
```diff
@@ -30,6 +37,225 @@ describe("sync/deployments", () => {
...
+					n.createdAt = clone.node.createdAt;
+					n.updatedAt = clone.node.updatedAt;
+				});
+				return clone;
+			});
+		};
+
+		const nockFetchingDeploymentgPagesGraphQL = (query, cursor, deployments) => {
+			githubNock.post("/graphql", { query, variables: { owner: repoSyncState.repoOwner, repo: repoSyncState.repoName, per_page: PAGE_SIZE__TWO_ITEMS, cursor } })
+				.query(true).reply(200, { "data": { "repository": { "deployments": { "edges": deployments } } } });
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:175]: quotations -> :killitwithfire:

File: src/sync/deployment.test.ts [hunk 13]
```diff
@@ -30,6 +37,225 @@ describe("sync/deployments", () => {
...
+						statuses_url: "random",
+						repository_url: "random",
+						transient_environment: false,
+						production_environmen: true
+					})));
+			});
+		};
+
+		const nockDeploymentStatusApi = (deployments, repeatTimes) => {
+			"*".repeat(repeatTimes).split("").forEach(() => {
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:202]: performance improvement: Array.from({ length: repeatTimes }).forEach(() => {

File: src/sync/deployment.test.ts [hunk 14]
```diff
@@ -30,6 +37,225 @@ describe("sync/deployments", () => {
...
+						.reply(200, [
+							{ id: idx * 1000 + 1, state: "inactive", description: "random", environment: item.node.environment },
+							{ id: idx * 1000 + 2, state: "success", description: "random", environment: item.node.environment }
+						]);
+				});
+			});
+		};
+
+		const nockDeploymentCommitGetApi = (deployments, repeatTimes) => {
+			"*".repeat(repeatTimes).split("").forEach(() => {
```
Comment (rachellerathbone) [inline src/sync/deployment.test.ts:214]: Samesies: Array.from({ length: repeatTimes }).forEach(() => {

File: src/sync/deployment.test.ts (hunk 7 above)
Comment (gxueatlassian) [reply src/sync/deployment.test.ts]: Nice one!! I wasn't aware of this `Array.from`. Was only looking at the `Array()` constructor. TIL. 👍

**Review by rachellerathbone — APPROVED** [review]

**Comment by rachellerathbone** [comment]
Looking forward to this going in. Hopefully we see some improvements in deployments :blobpray:

File: src/sync/deployment.ts [hunk 15]
```diff
@@ -49,12 +72,48 @@ const getTransformedDeployments = async (deployments, gitHubInstallationClient:
...
+	return (transformedDeployments
+		.filter(deployment => !!deployment) as JiraDeploymentBulkSubmitData[])
 		.map(deployment => deployment.deployments)
 		.flat();
 };
 
+const saveDeploymentsForLaterUse = async (deployments: FetchDeploymentResponse["deployments"], gitHubBaseUrl: string, logger: Logger) => {
+	try {
+
+		const successDeployments: FetchDeploymentResponse["deployments"] = deployments.filter(d => (d.statuses?.nodes.some(n => n.state === "SUCCESS")));
```
Comment (joshkay10) [inline src/sync/deployment.ts:84]: /nit Your can drop the type for `const successDeployments: FetchDeploymentResponse["deployments"]` since its inferred from the filter response.

File: src/sync/deployment.ts [hunk 16]
```diff
@@ -66,7 +125,12 @@ export const getDeploymentTask = async (
 ) => {
 	logger.debug("Syncing Deployments: started");
 
-	const { edges, deployments } = await fetchDeployments(gitHubInstallationClient, repository, cursor, perPage);
+	const { edges, deployments, extraDeployments } = await fetchDeployments(jiraHost, gitHubInstallationClient, repository, logger, cursor, perPage);
+
+	const useDyanmoForBackfill = await booleanFlag(BooleanFlags.USE_DYNAMODB_FOR_DEPLOYMENT_BACKFILL, jiraHost);
+	if (useDyanmoForBackfill) {
```
Comment (joshkay10) [inline src/sync/deployment.ts:131]: `useDyanmoForBackfill` to `useDynamoForBackfill`

File: src/github/client/github-installation-client.ts [hunk 17]
```diff
@@ -355,8 +356,12 @@ export class GitHubInstallationClient extends GitHubClient {
 		return response?.data?.data;
 	}
 
-	public async getDeploymentsPage(owner: string, repoName: string, perPage?: number, cursor?: string | number): Promise<getDeploymentsResponse> {
-		const response = await this.graphql<getDeploymentsResponse>(getDeploymentsQuery,
+	public async getDeploymentsPage(jiraHost: string, owner: string, repoName: string, perPage?: number, cursor?: string | number): Promise<getDeploymentsResponse> {
```
Comment (joshkay10) [inline src/github/client/github-installation-client.ts:359]: Would the cursor always be a string in this case?

File: src/sync/deployment.ts (hunk 15 above)
Comment (gxueatlassian) [reply src/sync/deployment.ts:84]: Good call, it must be my glitch on my vim ts settigs that doesn't show up correctly, so I added the extra types.
Sorry since this is already deployed, I am leaving this as it is right now. I may remember to remove it in next PR.

File: src/sync/deployment.ts (hunk 16 above)
Comment (gxueatlassian) [reply src/sync/deployment.ts:131]: 😞 crap. I'll leave this right now, hopefully the ff will be removed once 100% and this var will be removed in the followup PR.

File: src/github/client/github-installation-client.ts (hunk 17 above)
Comment (gxueatlassian) [reply src/github/client/github-installation-client.ts:359]: Yup I think so, 
1. The typing I believe it is because we use rest api and was page no before? So typing to be backward compatibility?
2. And can confirm all deploymentCursor is string (not number as string) in prod db. I'll remember to remove this on ff cleanup.
